GH-43070: [C++][Parquet] Check for valid ciphertext length to prevent segfault

adamreeve opened the pull request:
### Rationale for this change

See #43070

### What changes are included in this PR?

Checks that the ciphertext length is at least enough to hold the length (if written), nonce and GCM tag for the GCM cipher type.

Also enforces that the input ciphertext length parameter is provided (is > 0) and verifies that the ciphertext size read from the file isn't going to cause reads beyond the end of the ciphertext buffer.

### Are these changes tested?

Yes I've added new unit tests for this.

### Are there any user-facing changes?

No
* GitHub Issue: #43070

adamreeve commented:
Ah it looks like all the Windows builds are failing as I'm using non-exported classes in the new tests. Would it make sense to add `PARQUET_EXPORT` to these, or should I not be testing these internal classes? I think it would be quite difficult to add tests for this change at a higher level.

mapleFU commented:
The code LGTM, but I'm not familiar with decrypt module. So @wgtmac @pitrou for help

pitrou commented on cpp/src/parquet/CMakeLists.txt:
```diff
@@ -414,6 +414,7 @@ if(PARQUET_REQUIRE_ENCRYPTION)
                    encryption/test_encryption_util.cc)
   add_parquet_test(encryption-key-management-test
                    SOURCES
+                   encryption/encryption_internal_test.cc
```
I would add this to `encryption-test` rather, since this isn't related to key management?

adamreeve replied:
Yes sorry I missed that there was a separate `encryption-test` target

pitrou commented on cpp/src/parquet/encryption/encryption_internal.cc:
```diff
@@ -463,10 +463,13 @@ int AesDecryptor::AesDecryptorImpl::GcmDecrypt(const uint8_t* ciphertext,
+    ss << "Invalid ciphertext length " << ciphertext_len << ". Expected at least "
+       << length_buffer_length_ + kNonceLength + kGcmTagLength << "\n";
+    throw ParquetException(ss.str());
   }
```
Looking at the existing code in these methods, several things stand out:
1) we don't validate ciphertext length before fetching the 4 bytes encoding the actual length (a corrupt file could perhaps have a ciphertext length < 4?)
2) use of raw C arrays instead of `std::array<uint8_t>` for example
3) why is the `ciphertext_len` argument optional in this API? this looks fickle and error-prone.
4) the part that extracts and validates the actual length is duplicated in the two methods

I would suggest we take the opportunity and refactor this into a cleaner and less error-prone implementation. In particular, the GCM and CTR-specific methods should probably have a mandatory ciphertext length, and would not have to bother with reading the length bytes.

@ggershinsky @thamht4190 Do we have an explanation for the very odd choices here?

adamreeve replied:
Yeah I didn't dig in too deep to find why the `ciphertext_len` is optional, it would be nice if that could be mandatory. But if that's not possible we should at least be able to provide the size of the buffer that the ciphertext is being read from to ensure that `ciphertext_len` isn't greater than this.

adamreeve replied:
It does look like we don't always know the length of the ciphertext but sometimes just an upper bound that's used to allocate the buffer. I found an example of this when reading the bloom filter header:
https://github.com/apache/arrow/blob/e615a3034ed70c9bdbc0b74882f7345c0c2b8912/cpp/src/parquet/bloom_filter.cc#L116-L121

I'm thinking that rather than having separate arguments like `ciphertext_buffer_len` (required) and `ciphertext_expected_len` (optional), it's probably fine to make `ciphertext_len` required and mean the size of the buffer, so we would validate that the actual length is <= this after accounting for the 4 byte length rather than enforcing an exact match. Does that seem reasonable?

pitrou commented:
Another potential weak point is this:
https://github.com/apache/arrow/blob/1da71ba41ce8844eb89935bce44e14d966a262e6/cpp/src/parquet/thrift_internal.h#L415-L417

and of course this part where we totally ignore the physical buffer length, letting the `Decrypt` function happily read past the end of the buffer:
https://github.com/apache/arrow/blob/1da71ba41ce8844eb89935bce44e14d966a262e6/cpp/src/parquet/thrift_internal.h#L419-L420

All in all, this warrants a refactor for sanity and robustness.

adamreeve commented:
@pitrou, I think I've addressed all of your comments now thank you
* I've added extra validation of the buffer sizes in various places
* I've switched from raw pointers to `arrow::util::span`
* The ciphertext size is no longer optional, although it's no longer checked for an exact match, the actual ciphertext length might be less than the buffer size
* I've pulled out the length reading and validation into a shared method
* I've refactored how the plaintext/ciphertext length conversions are handled by adding methods for these rather than adding or subtracting the size delta in consumer code

I haven't touched the `AesEncryptor` API at all, but it would probably make sense to follow up after this to at least change that to use `arrow::util::span` too for consistency.

pitrou commented:
@adamreeve I just want to let you know that I'm currently sick and may not be able to review this before the next week. Thanks for doing this!

adamreeve commented:
OK no problem, thanks for letting me know, and I hope you're feeling better soon.

raulcd commented:
@mapleFU @wgtmac is there any possibility you could take a look on this? Otherwise this fix will have to miss the 17.0.0 release unfortunately

wgtmac commented:
I'll try to review this today.

mapleFU commented on cpp/src/parquet/encryption/read_configurations_test.cc:
```diff
@@ -22,6 +22,7 @@
 
 #include "arrow/io/file.h"
 #include "arrow/testing/gtest_compat.h"
+#include "arrow/util/config.h"
```
Ever used?

adamreeve replied:
Yes this is needed so that `ARROW_WITH_SNAPPY` is defined, otherwise the tests below are always skipped. This is a bit unrelated to this change but I noticed this problem when running the tests locally.

mapleFU commented on cpp/src/parquet/encryption/encryption_internal.h:
```diff
@@ -44,7 +45,7 @@ constexpr int8_t kBloomFilterHeader = 8;
 
 /// Performs AES encryption operations with GCM or CTR ciphers.
-class AesEncryptor {
+class PARQUET_EXPORT AesEncryptor {
```
Emm why exporting this?

adamreeve replied:
Yeah the tests fail to build on Windows without this. I could maybe have tested via the `parquet::encryption::Decryptor` class which is already exported, but the constructor for that needs an `AesDecryptor` so I would still have needed to export these classes anyway, or provide a new method to work around that.

mapleFU reviewed (commented):
The style looks ok but I'm not familiar with encryption

wgtmac commented on cpp/src/parquet/encryption/encryption_internal.cc:
```diff
@@ -438,56 +462,102 @@ std::shared_ptr<AesDecryptor> AesDecryptor::Make(
-                                               int aad_len, uint8_t* plaintext) {
-  int len;
-  int plaintext_len;
+int AesDecryptor::PlaintextLength(int ciphertext_len) const {
```
Should we replace `int` with `int32_t` to be more portable? Same for functions below. Of course this can be a followup change.

adamreeve replied:
Yes it would make sense to do this as a follow up to avoid changing too much in this PR, I've made https://github.com/apache/arrow/issues/43141 for this

wgtmac commented on cpp/src/parquet/encryption/encryption_internal.cc:
```diff
@@ -438,56 +462,102 @@ std::shared_ptr<AesDecryptor> AesDecryptor::Make(
+int AesDecryptor::AesDecryptorImpl::GetCiphertextLength(
+    ::arrow::util::span<const uint8_t> ciphertext) const {
   if (length_buffer_length_ > 0) {
+    if (ciphertext.size() < static_cast<size_t>(length_buffer_length_)) {
```
It took me a while to understand this line. Perhaps it is better to explicitly use `kBufferSizeLength` here as line 484 to 486 have assumed this length is 4 bytes.

adamreeve replied:
Yes good point, that also confused me a bit. I've changed that and added a comment to hopefully make this more readable

wgtmac commented on cpp/src/parquet/encryption/encryption_internal.cc:
```diff
@@ -438,56 +462,102 @@ std::shared_ptr<AesDecryptor> AesDecryptor::Make(
+    if (ciphertext.size() < static_cast<size_t>(length_buffer_length_)) {
+      std::stringstream ss;
+      ss << "Ciphertext buffer length " << ciphertext.size()
+         << " is insufficient to read the ciphertext length";
```
nit: include the length (4 bytes) in the error message.

adamreeve replied:
:+1: fixed

wgtmac reviewed (commented):
Generally LGTM. Thanks for the fix!

adamreeve commented on cpp/src/parquet/encryption/encryption.h:
```diff
@@ -89,6 +89,14 @@ inline const uint8_t* str2bytes(const std::string& str) {
   return reinterpret_cast<const uint8_t*>(cbytes);
 }
 
+inline ::arrow::util::span<const uint8_t> str2span(const std::string& str) {
```
You can already construct a span from a string but that creates a `span<const char>`. Converting to a `uint8_t` span might be more specific to the encryption use case so I'm not sure about this.

adamreeve commented on cpp/src/parquet/encryption/encryption_internal_test.cc:
```diff
@@ -0,0 +1,144 @@
+    std::vector<uint8_t> ciphertext(expected_ciphertext_len, '\0');
+
+    int ciphertext_length =
+        encryptor.Encrypt(str2bytes(plain_text_), static_cast<int>(plain_text_.size()),
```
Yes I've just made https://github.com/apache/arrow/issues/43142 for this, and I will follow up after this PR

adamreeve commented on cpp/src/parquet/encryption/encryption_internal.cc:
```diff
@@ -325,25 +325,43 @@ class AesDecryptor::AesDecryptorImpl {
+  }
+
+  [[nodiscard]] int CiphertextLength(int plaintext_len) const {
+    return plaintext_len + ciphertext_size_delta_;
```
:+1: I've added a check for this

adamreeve commented on cpp/src/parquet/encryption/encryption_internal.cc:
```diff
@@ -315,8 +315,10 @@ class AesDecryptor::AesDecryptorImpl {
 
-  int Decrypt(const uint8_t* ciphertext, int ciphertext_len, const uint8_t* key,
-              int key_len, const uint8_t* aad, int aad_len, uint8_t* plaintext);
+  int Decrypt(::arrow::util::span<const uint8_t> ciphertext,
```
Good point, I've done that now

adamreeve commented on cpp/src/parquet/encryption/encryption_internal.cc:
```diff
@@ -438,56 +462,102 @@ std::shared_ptr<AesDecryptor> AesDecryptor::Make(
   } else {
-    if (ciphertext_len == 0) {
-      throw ParquetException("Zero ciphertext length");
+    if (ciphertext.size() > static_cast<size_t>(std::numeric_limits<int>::max())) {
```
:+1: fixed

wgtmac reviewed (approved):
+1

Thanks!

wgtmac commented:
CI failures are unrelated. I will merge it shortly.

wgtmac commented:
@raulcd Please feel free to port it to maint-17.0.0

mapleFU commented:
(Gang says he has api error when run merge script so I merged this, lol)

raulcd commented:
Thanks both for jumping in on the review and thanks @adamreeve for the PR

pitrou commented on cpp/src/parquet/encryption/encryption_internal.cc:
```diff
@@ -325,25 +326,46 @@ class AesDecryptor::AesDecryptorImpl {
+      ss << "Negative plaintext length " << plaintext_len;
+      throw ParquetException(ss.str());
+    }
+    return plaintext_len + ciphertext_size_delta_;
```
We should ideally check for signed addition overflow here.

adamreeve replied:
Good point. Is it ok to add this to the changes in #43195, or should I make a separate PR for these follow ups?

pitrou replied:
IMHO you can add them to the followup PR.

adamreeve replied:
I've added this check in https://github.com/apache/arrow/pull/43195/commits/fb63bcc7456a401b81f5fbacd6b728550c0d788e

pitrou commented on cpp/src/parquet/encryption/encryption_internal.cc:
```diff
@@ -438,56 +460,105 @@ std::shared_ptr<AesDecryptor> AesDecryptor::Make(
-    if (ciphertext_len > 0 &&
-        ciphertext_len != (written_ciphertext_len + length_buffer_length_)) {
-      throw ParquetException("Wrong ciphertext length");
+    if (written_ciphertext_len < 0) {
```
I wonder if the compiler won't elide this, as `((ciphertext[3] & 0xff) << 24)` becoming negative implies signed integer overflow which is undefined behavior. @felipecrv What do you think?

felipecrv replied:
This check is weird. `written_ciphertext_len` should be `uint32_t`, then checking that `written_ciphertext_len` is not greater than `std::numeric_limits<int32_t>::max()` is enough. In other words: the encoded value is never negative, but it can be too big.

pitrou replied:
`ciphertext[3] & 0xff` is weird as well, since `ciphertext[3]` is a uint8, thus in [0,255] already.
Perhaps the proper check should simply be `ciphertext[3] < 0x80u`?

adamreeve replied:
Hmm yeah the `& 0xff`s are all redundant, maybe it ended up written this way to be similar to the code for writing the length. Rather than just checking the length isn't negative, we should also check that `written_ciphertext_len + length_buffer_length_` doesn't overflow, so I think it would probably be simplest to read as `uint32_t`.

adamreeve replied:
I've changed this check as part of #43195, in this commit: https://github.com/apache/arrow/pull/43195/commits/b81cafe1a74cd821fde4d2e58481359149eaf233

pitrou commented on cpp/src/parquet/encryption/encryption_internal.cc:
```diff
@@ -438,56 +460,105 @@ std::shared_ptr<AesDecryptor> AesDecryptor::Make(
+  int len;
+  int plaintext_len;
+
+  uint8_t tag[kGcmTagLength];
+  memset(tag, 0, kGcmTagLength);
```
Or more idiomatically:
```c++
  std::array<uint8_t, kGcmTagLength> tag{};
```

adamreeve replied:
I've changed this as well as the other uses of C style arrays in https://github.com/apache/arrow/pull/43195/commits/ff31c7b23484ef8194ec7e098303d44b92e5b8da

pitrou commented on cpp/src/parquet/thrift_internal.h:
```diff
@@ -411,17 +411,21 @@ class ThriftDeserializer {
-          AllocateBuffer(decryptor->pool(),
-                         static_cast<int64_t>(clen - decryptor->CiphertextSizeDelta())));
-      const uint8_t* cipher_buf = buf;
+      auto decrypted_buffer = std::static_pointer_cast<ResizableBuffer>(AllocateBuffer(
```
Unrelated, but since we're not attempting to resize the buffer, the cast to `ResizableBuffer` should be superfluous?

adamreeve replied:
`AllocateBuffer` here is `::parquet::AllocateBuffer` which returns a `std::shared_ptr<ResizableBuffer>`, rather than `::arrow::AllocateBuffer`, so it looks like the cast would be unnecessary even if we were resizing it:
https://github.com/apache/arrow/blob/031497ddab5ce497c49b87d7125bf5c01bf80f48/cpp/src/parquet/platform.cc#L36-L37

pitrou replied:
Hmm, I see. We can try removing it at some point then, as the code currently looks weird.

adamreeve replied:
I've removed this as well as another occurrence of the same behaviour in https://github.com/apache/arrow/pull/43195/commits/f4a7721d13317c013b2b17aa099bb060c9fbe046

pitrou reviewed (commented):
Belated review, sorry.
